src/kaleidoscope/driver/Bootloader.h

```diff
@@ -0,0 +1,124 @@
+/* -*- mode: c++ -*-
+ * kaleidoscope::driver::bootloader -- Various bootloader drivers
+ * Copyright (C) 2019  Keyboard.io, Inc
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <avr/wdt.h>
+
+namespace kaleidoscope {
+namespace driver {
+namespace bootloader {
+
+class Caterina {
+ public:
+  Caterina() {}
+
+  static inline void resetDevice() {
+    // Set the magic bits to get a Caterina-based device
+    // to reboot into the bootloader and stay there, rather
+    // than run move onward
+    //
+    // These values are the same as those defined in
+    // Caterina.c
+
+    uint16_t bootKey = 0x7777;
+    uint16_t *const bootKeyPtr = reinterpret_cast<uint16_t *>(0x0800);
+
+    // Stash the magic key
+    *bootKeyPtr = bootKey;
+
+    // Set a watchdog timer
+    wdt_enable(WDTO_120MS);
+
+    while (1) {} // This infinite loop ensures nothing else
+    // happens before the watchdog reboots us
+  }
+};
+
+class HalfKay {
+ public:
+  HalfKay() {}
+
+  static void resetDevice() {
+    cli();
+    UDCON = 1;
+    USBCON = (1 << FRZCLK);
+    UCSR1B = 0;
+    _delay_ms(5);
+
+    EIMSK = 0;
+    PCICR = 0;
+    SPCR = 0;
+    ACSR = 0;
+    EECR = 0;
+    ADCSRA = 0;
+    TIMSK0 = 0;
+    TIMSK1 = 0;
+    TIMSK3 = 0;
+    TIMSK4 = 0;
+    UCSR1B = 0;
+    TWCR = 0;
+    DDRB = 0;
+    DDRC = 0;
+    DDRD = 0;
+    DDRE = 0;
+    DDRF = 0;
+    TWCR = 0;
+    PORTB = 0;
+    PORTC = 0;
+    PORTD = 0;
+    PORTE = 0;
+    PORTF = 0;
+    asm volatile("jmp 0x7E00");
+  }
+};
+
+#ifdef KALEIDOSCOPE_BOOTLOADER_DFU
+
+#define BOOTLOADER_RESET_KEY 0xB007B007
+static uint32_t reset_key  __attribute__((section(".noinit")));
+
+/*
+ * This function runs before main(), and jumps to the bootloader after a reset
+ * initiated by .resetDevice().
+ */
+static void _bootloader_jump_after_watchdog_reset()
+__attribute__((used, naked, section(".init3")));
+static void _bootloader_jump_after_watchdog_reset() {
+  if ((MCUSR & (1 << WDRF)) && reset_key == BOOTLOADER_RESET_KEY) {
+    reset_key = 0;
+
+    ((void (*)(void))(((FLASHEND + 1L) - 4096) >> 1))();
+  }
+}
+
+class DFU {
+ public:
+  DFU() {}
+
+  static inline void resetDevice() {
+    reset_key = BOOTLOADER_RESET_KEY;
+    wdt_enable(WDTO_250MS);
+    for (;;);
```

Comment:
Another style of infinite loop here, in the same file.

Reply:
Good catch. These were all copied from different - usually C - sources, hence the different styles. I'll update the PR to use `while (true)` instead. Thanks!